url29ff2e53f3_nicolas3470_Settlers_of_Catan_AI_tgz-pJ8-robotGameJ8/src/soc/client/SOCBoardPanel.java

```diff
@@ -0,0 +1,5144 @@
+    public int[] scaleCopyToActualY(int[] yorig)
+    {
+        int[] ys = new int[yorig.length];
+        for (int i = yorig.length - 1; i >= 0; --i)
+            ys[i] = (int) ((yorig[i] * (long) scaledPanelY) / panelMinBH);
+        return ys;
+    }
+
+    /**
+     * Copy and rotate this array of y-coordinates, optionally also rescaling.
+     * Rotates internal to actual (clockwise):  P'=(width-y, x)
+     * @param yorig Array to copy and rotate, not null
+     * @param width Width to rotate against
+     * @param rescale Should we also rescale, same formula as {@link #scaleCopyToActualX(int[])}?
+     * @return Rotated copy of <tt>yorig</tt> for use as x-coordinates
+     * @since 1.1.08
+     */
+    public int[] rotateScaleCopyYToActualX(final int[] yorig, final int width, final boolean rescale)
+    {
+        int[] xr = new int[yorig.length];
+        for (int i = yorig.length - 1; i >= 0; --i)
+            xr[i] = width - yorig[i];
+        if (rescale)
+            for (int i = yorig.length - 1; i >= 0; --i)
+                xr[i] = (int) ((xr[i] * (long) scaledPanelX) / panelMinBW);
+        return xr;
+    }
+
+    /**
+     * Redraw the board using double buffering. Don't call this directly, use
+     * {@link Component#repaint()} instead.
+     */
+    public void paint(Graphics g)
+    {
+        try {
+        if (buffer == null)
+        {
+            buffer = this.createImage(scaledPanelX, scaledPanelY);
+        }
+        drawBoard(buffer.getGraphics());
+        if (hoverTip.isVisible())
+            hoverTip.paint(buffer.getGraphics());
+        buffer.flush();
+        g.drawImage(buffer, 0, 0, this);
+
+        } catch (Throwable th) {
+            playerInterface.chatPrintStackTrace(th);
+        }
+    }
+
+    /**
+     * Overriden so the peer isn't painted, which clears background. Don't call
+     * this directly, use {@link Component#repaint()} instead.
+     */
+    public void update(Graphics g)
+    {
+        paint(g);
+    }
+
+    /**
+     * draw a board tile
+     * @param g       graphics
+     * @param hexNum  hex number (0-36)
+     */
+    private final void drawHex(Graphics g, int hexNum)
+    {
+        drawHex(g, hexX[hexNum], hexY[hexNum], board.getHexLayout()[hexNum], hexNum);
+    }
+
+    /**
+     * draw a given type of board tile at a given location.
+     * Use <tt>hexType</tt> to determine the hex graphic and possibly a port overlay graphic.
+     * When <tt>hexNum</tt> is -1, nothing else is drawn.
+     * Otherwise, also draw the dice number for this hex (if any).
+     *
+     * @param g      graphics
+     * @param x      board-graphics x-coordinate to draw at
+     * @param y      board-graphics y-coordinate to draw at
+     * @param hexType hex number, as in {@link SOCBoard#getHexLayout()}
+     * @param hexNum  hex number (0-36), or -1 if this isn't a valid hex number
+     * @since 1.1.08
+     */
+    private final void drawHex(Graphics g, int x, int y, int hexType, int hexNum)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 34. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 1847-1909 into a separate method.